### PR TITLE
[tx] Implement context parallelism in tx with ring attention using `ppermute`

### PR DESCRIPTION
**Pull request by tanmaysachan**
Addresses https://github.com/NovaSky-AI/SkyRL/issues/1056

This PR implements Context Parallelism (CP).
CP is now an axis alongside fsdp/ep/tp, with parity coverage against CP=1.

[AI wrote the description]

## What’s included
- Added context_parallel_size to model/backend configs.
- Added CLI support via --cp-size in tx/run/train.py.
- Extended mesh construction to 4D: ("fsdp", "ep", "tp", "cp").
- Implemented CP-aware attention path:
    - ring-style KV exchange via ppermute
    - streaming softmax accumulation across CP shards
    - CP-aware default token positions.
- Updated model attention callsites (Qwen3/Llama3/DeepSeek) to pass positions and use CP-aware attention helper.
- Updated KV-cache layer updates for CP ownership/local position handling.
- Updated sequence-length bucketing utility to support CP-aware padding (round_up_seq_len(..., cp=...)).
- Refactored JAX backend train/sample paths for CP:
    - shard-map based model pass with CP partition specs
    - CP-specific sample pass with cached compilation
    - unified generate() path with optional decode runner override.
- Added/updated tests for CP config and runtime behavior:
    - CP forward/backward execution
    - CP sample execution
    - CP parity (sample outputs, prompt logprobs, forward loss/logprobs, backward gradients)
    - attention test updates for positions API.

## Validation
- Added targeted CP tests in tests/tinker/test_jax_backend.py.
- Updated tests/models/test_qwen3_config.py and tests/gpu/test_attention.py for new CP/positions behavior.

Tested over multi-cpu configs.

## Tasks
- [x] Further simplify decode/prefill runner interfaces and reduce remaining wrapper boilerplate.

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1149" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by tanmaysachan** [comment]
PR ready for review.

One ugliness is the wrapping of prefill_and_decode in a shard_map in JaxBackendImpl. Do let me know if a cleaner design is preferred

File: skyrl-tx/tx/run/train.py
[the remark this answers is not in the document]
Comment (tanmaysachan) [reply skyrl-tx/tx/run/train.py:1]: Good bot

File: skyrl-tx/tx/utils/generator.py [hunk 1]
```diff
@@ -286,21 +297,27 @@ def generate(
...
                 Required if any sampling_params has stop_strings set.
+            decode_runner: Optional callable override for `_prefill_and_decode`
+                (for example, CP shard_map path).
 
         Returns:
             GenerateOutput containing generated_ids, stop_reasons, and optionally logprobs.
         """
         batch_size, prompt_length = input_ids.shape
+        if positions is None:
+            positions = jnp.arange(prompt_length, dtype=jnp.int32)[None, :]
```
[the remark this answers is not in the document]
Comment (tanmaysachan) [reply skyrl-tx/tx/utils/generator.py:320]: The axis_index fpr cp might not exist. default_positions is used when we are under a shard_map

File: skyrl-tx/tx/tinker/backends/jax.py [hunk 2]
```diff
@@ -810,26 +881,26 @@ def sample(
                     all_sampling_params[batch_start:batch_end], max_batch_size, fill=all_sampling_params[batch_start]
                 )
 
-                # Pad sequences to same length within the batch to minimize memory usage.
-                # Also bin it so the JIT has to compile fewer kernels.
                 max_len = round_up_seq_len(max((len(seq) for seq in batch_prompts), default=0))
                 input_ids = pad_batch(batch_prompts, max_len, np.int32)
                 attention_mask = pad_batch([[1] * len(seq) for seq in batch_prompts], max_len, np.int32)
+                positions = np.arange(max_len, dtype=np.int32)[None, :]
```
[the remark this answers is not in the document]
Comment (tanmaysachan) [reply skyrl-tx/tx/tinker/backends/jax.py:887]: Same as the other - we need to be under a shard_map for cp axis_index to exist. This positions can also optionally be left as None to be filled by the generator.

File: skyrl-tx/tx/utils/generator.py
[the remark this answers is not in the document]
Comment (tanmaysachan) [reply skyrl-tx/tx/utils/generator.py:1]: Yeah, finding a way to do this cleanly
